### PR TITLE
Fix race condition in SolutionChecksumUpdater work queue

### LINKED ISSUE #46255: The UpdaterService test is flaky
### Builds
|Build|Pull Request | Test Failure Count|
| --- | --- | --- |
|[#741836](https://dev.azure.com/dnceng/public/_build/results?buildId=741836)|Rolling|1|
|[#742047](https://dev.azure.com/dnceng/public/_build/results?buildId=742047)|#46210|1|
|[#742073](https://dev.azure.com/dnceng/public/_build/results?buildId=742073)|#46229|1|
|[#742078](https://dev.azure.com/dnceng/public/_build/results?buildId=742078)|Rolling|1|
|[#742108](https://dev.azure.com/dnceng/public/_build/results?buildId=742108)|#46233|1|
|[#742242](https://dev.azure.com/dnceng/public/_build/results?buildId=742242)|Rolling|2|
|[#742430](https://dev.azure.com/dnceng/public/_build/results?buildId=742430)|#46234|1|
|[#742466](https://dev.azure.com/dnceng/public/_build/results?buildId=742466)|#46026|1|
|[#742550](https://dev.azure.com/dnceng/public/_build/results?buildId=742550)|Rolling|1|
|[#742551](https://dev.azure.com/dnceng/public/_build/results?buildId=742551)|Rolling|1|
|[#742560](https://dev.azure.com/dnceng/public/_build/results?buildId=742560)|#44447|1|
|[#742614](https://dev.azure.com/dnceng/public/_build/results?buildId=742614)|#45348|2|
|[#742716](https://dev.azure.com/dnceng/public/_build/results?buildId=742716)|#45348|1|
|[#742765](https://dev.azure.com/dnceng/public/_build/results?buildId=742765)|Rolling|1|
|[#742784](https://dev.azure.com/dnceng/public/_build/results?buildId=742784)|#45993|2|
|[#743055](https://dev.azure.com/dnceng/public/_build/results?buildId=743055)|#44447|1|
### Configurations
- Windows Desktop Debug Test32
- Windows Desktop Debug Test64
- Windows Desktop Release Test32
- Windows Desktop Release Test64
- Windows Desktop Spanish

### PR DESCRIPTION
**Pull request by sharwell**
Fixes #46255

### CODE REVIEW

File: src/VisualStudio/Core/Def/Implementation/Remote/SolutionChecksumUpdater.cs [hunk 1]
```diff
@@ -52,8 +53,9 @@ protected override async Task ExecuteAsync()
         {
             lock (_gate)
             {
-                _lastToken?.Dispose();
-                _lastToken = null;
+                Contract.ThrowIfFalse(_currentToken is object);
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/Remote/SolutionChecksumUpdater.cs:56]: Contract.ThrowIfNull?

File: src/VisualStudio/Core/Def/Implementation/Remote/SolutionChecksumUpdater.cs [hunk 2]
```diff
@@ -102,17 +111,12 @@ private void OnWorkspaceChanged(object sender, WorkspaceChangeEventArgs e)
         private void EnqueueChecksumUpdate()
         {
             // event will raised sequencially. no concurrency on this handler
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/Remote/SolutionChecksumUpdater.cs:113]:
```suggestion
            // event will raised sequentially. no concurrency on this handler
```

File: src/VisualStudio/Core/Def/Implementation/Remote/SolutionChecksumUpdater.cs [hunk 3]
```diff
@@ -73,8 +75,15 @@ protected override void PauseOnGlobalOperation()
...
-        protected override Task WaitAsync(CancellationToken cancellationToken)
-            => _event.WaitAsync(cancellationToken);
+        protected override async Task WaitAsync(CancellationToken cancellationToken)
+        {
+            var currentToken = await _workQueue.DequeueAsync(cancellationToken).ConfigureAwait(false);
+            lock (_gate)
+            {
+                Contract.ThrowIfFalse(_currentToken is null);
+                _currentToken = currentToken;
+            }
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/Remote/SolutionChecksumUpdater.cs:85]: tbh, i've stared at thsi a bunch, and i have no feeling on what's happening here or why the new code is more correct.  Can you write a non-brief explanation of what this *class* is trying to accomplish, how tehse two fields play into it, and why the new approach is 'correct'?

Comment (sharwell) [reply src/VisualStudio/Core/Def/Implementation/Remote/SolutionChecksumUpdater.cs:85]: https://github.com/dotnet/roslyn/issues/46255#issuecomment-664043259

Comment (sharwell) [reply src/VisualStudio/Core/Def/Implementation/Remote/SolutionChecksumUpdater.cs:85]: image

File: src/VisualStudio/Core/Def/Implementation/Remote/SolutionChecksumUpdater.cs (hunk 2 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Implementation/Remote/SolutionChecksumUpdater.cs:113]: I don't currently have my spell checking extension installed. I'm good leaving this for a separate change focused on spelling and grammar.

**Pull request merged**